pkg/appgw/requestroutingrules.go

```diff
@@ -59,14 +59,50 @@ func (c *appGwConfigBuilder) RequestRoutingRules(cbCtx *ConfigBuilderContext) er
-	httpListenersMap := c.groupListenersByListenerIdentifier(c.getListeners(cbCtx))
+func (c *appGwConfigBuilder) getRules(cbCtx *ConfigBuilderContext) ([]n.ApplicationGatewayRequestRoutingRule, []n.ApplicationGatewayURLPathMap) {
+	httpListenersMap := c.groupListenersByListenerIdentifier(c.appGw.HTTPListeners)
+	var pathMap []n.ApplicationGatewayURLPathMap
+	var requestRoutingRules []n.ApplicationGatewayRequestRoutingRule
+	for listenerID, urlPathMap := range c.getPathMaps(cbCtx) {
+		httpListener := httpListenersMap[listenerID]
+		rule := n.ApplicationGatewayRequestRoutingRule{
+			Etag: to.StringPtr("*"),
+			Name: to.StringPtr(generateRequestRoutingRuleName(listenerID)),
+			ID:   to.StringPtr(c.appGwIdentifier.requestRoutingRuleID(generateRequestRoutingRuleName(listenerID))),
+			ApplicationGatewayRequestRoutingRulePropertiesFormat: &n.ApplicationGatewayRequestRoutingRulePropertiesFormat{
+				HTTPListener: &n.SubResource{ID: to.StringPtr(c.appGwIdentifier.listenerID(*httpListener.Name))},
+			},
+		}
+		glog.V(5).Infof("Binding rule %s to listener %s", *rule.Name, *httpListener.Name)
+		if len(*urlPathMap.PathRules) == 0 {
+			// Basic Rule, because we have no path-based rule
+			rule.RuleType = n.Basic
+			rule.RedirectConfiguration = urlPathMap.DefaultRedirectConfiguration
+
+			// We setup the default backend address pools and default backend HTTP settings only if
+			// this rule does not have an `ssl-redirect` configuration.
+			if rule.RedirectConfiguration == nil {
+				rule.BackendAddressPool = urlPathMap.DefaultBackendAddressPool
+				rule.BackendHTTPSettings = urlPathMap.DefaultBackendHTTPSettings
+			}
+		} else {
+			// Path-based Rule
+			rule.RuleType = n.PathBasedRouting
+			rule.URLPathMap = &n.SubResource{ID: to.StringPtr(c.appGwIdentifier.urlPathMapID(*urlPathMap.Name))}
+			pathMap = append(pathMap, *urlPathMap)
+		}
+		requestRoutingRules = append(requestRoutingRules, rule)
+	}
+	return requestRoutingRules, pathMap
+}
+
+func (c *appGwConfigBuilder) getPathMaps(cbCtx *ConfigBuilderContext) map[listenerIdentifier]*n.ApplicationGatewayURLPathMap {
+	defaultAddressPoolID := to.StringPtr(c.appGwIdentifier.addressPoolID(defaultBackendAddressPoolName))
+	defaultHTTPSettingsID := to.StringPtr(c.appGwIdentifier.httpSettingsID(defaultBackendHTTPSettingsName))
+	httpListenersMap := c.groupListenersByListenerIdentifier(c.appGw.HTTPListeners)
 	urlPathMaps := make(map[listenerIdentifier]*n.ApplicationGatewayURLPathMap)
-	backendPools := c.newBackendPoolMap(cbCtx)
-	_, backendHTTPSettingsMap, _, _ := c.getBackendsAndSettingsMap(cbCtx)
-	for _, ingress := range cbCtx.IngressList {
-		defaultAddressPoolID := c.appGwIdentifier.addressPoolID(defaultBackendAddressPoolName)
-		defaultHTTPSettingsID := c.appGwIdentifier.httpSettingsID(defaultBackendHTTPSettingsName)
+	for ingressIdx := range cbCtx.IngressList {
+		ingress := cbCtx.IngressList[ingressIdx]
```

Comment:
Why do we change for `for _, ingress := range cbCtx.IngressList` to `for ingressIdx := range cbCtx.IngressList` if we only use `ingressIdx` to get the `ingress`?

Reply:
This because we using the rule pointer to get the backendID from the map. This requires the actual objects address rather than a copies address.

Reply:
https://github.com/Azure/application-gateway-kubernetes-ingress/pull/422/files#diff-9f4ecc464102fb9e5a7063f9a97115a3R240